pandas/core/indexing.py

```diff
@@ -663,17 +663,9 @@ def _ensure_listlike_indexer(self, key, axis=None, value=None):
             and not com.is_bool_indexer(key)
             and all(is_hashable(k) for k in key)
         ):
-            for i, k in enumerate(key):
-                if k not in self.obj:
-                    if value is None:
-                        self.obj[k] = np.nan
-                    elif is_array_like(value) and value.ndim == 2:
-                        # GH#37964 have to select columnwise in case of array
-                        self.obj[k] = value[:, i]
-                    elif is_list_like(value):
-                        self.obj[k] = value[i]
-                    else:
-                        self.obj[k] = value
+            keys = self.obj.columns.tolist()
+            keys.extend([k for k in key if k not in self.obj])
```

Comment:
why are you doing this with a list instead of Index.union?

Reply:
Haven't thoug about union. Thanks. Changed it.

Reply:
Gave us another slight performance improvement

```
In [16]: %timeit setitem(x, x_col, df)
1.03 ms ± 9.96 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```